notebooks/machine_learning/raw/tut8.ipynb

```diff
@@ -0,0 +1,479 @@
+      " [  4  91  15   0]\n",
+      " [  0  21  98  12]\n",
+      " [  0   0  10 113]]\n",
+      "\n",
+      "Normalized confusion matrix:\n",
+      "[0.95588235 0.04411765 0.         0.        ]\n",
+      "[0.03636364 0.82727273 0.13636364 0.        ]\n",
+      "[0.         0.16030534 0.7480916  0.09160305]\n",
+      "[0.         0.         0.08130081 0.91869919]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Calculate the confusion matrix itself\n",
+    "confusion = metrics.confusion_matrix(val_y, pred_y)\n",
+    "print(f\"Confusion matrix:\\n{confusion}\")\n",
+    "\n",
+    "\n",
+    "# Normalizing by the true label counts to get rates\n",
+    "print(f\"\\nNormalized confusion matrix:\")\n",
+    "for row in confusion:\n",
+    "    print(row / row.sum())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<img src=\"https://i.imgur.com/idD0k8y.png\" alt=\"example confusion matrix\" width=400px>\n",
+    "\n",
+    "The rows of the confusion matrix are the true class and the columns are the predicted class. The diagonal tells us how many of each class the model predicted correctly. The off-diagonals show where the model is making wrong predictions, where it is \"confused.\" For example, looking at the first column and second row, we classified four phones that were actually low cost as medium cost. We see for classes 0 and 3, the low cost and highest cost phones, our model works really well, above 90% accurate. However, our model is weaker for medium and high cost phones. Note that incorrect predictions are only between adjacent classes. The model doesn't confuse low cost and very high cost phones.\n",
+    "\n",
+    "The diagonal of the normalized confusion matrix gives us the **true positive rate** for each of the classes. This is the fraction of positive cases in the validation data the model predicted to be positive. This value is important because the accuracy metric can sometimes hide information you care about. Occasionally you'll have unbalanced classes where you have many more examples of one class compared to another. For example, it's typical that when you're trying to diagnose a disease with a machine learning model, most examples in your data won't have that disease. Your model could be very good at identifying when people don't have the disease (negative cases), but poor at identifying positive cases. \n",
+    "\n",
+    "To illustrate this, assume negative cases make up 90% of your data and the classifier identifies negative cases at 95% accuracy.  Positive cases make up the other 10% of the data but your classifier only has a 25% accuracy for these cases. The overall accuracy would then be $90\\% * 95\\% + 25\\% * 5\\% = 88\\%$. The accuracy looks high and if that was the only metric you looked at you might think your classifier is doing pretty well. However, the goal is to identify when people have this disease, so 25% accuracy there is clearly not sufficient."
```

Comment:
I found it easy for my mind to wander in this paragraph.

What do you think about making a ThoughtQuestion out of this topic?

So the user gets to actively think instead of passively reading about it?